src/lavinmq/amqp/queue/message_store.cr

```diff
@@ -399,33 +413,85 @@ module LavinMQ
-            next if deleted?(seg, pos)
+          next if mfile.size <= 4 # only got the schema version
+          # check if message count in segment is appended to the end
+          begin
+            count = read_msg_count(mfile)
+            @segment_msg_count[seg] = count
+            bytesize = mfile.size - 4 # minus schema version
+            acked = 0
+            if deleted = @deleted[seg]?
+              acked = deleted.size
+              deleted.each do |pos|
+                mfile.pos = pos
+                bytesize -= BytesMessage.skip(mfile)
+              end
+            end
             @bytesize += bytesize
-            @size += 1
-          rescue ex : IO::EOFError
-            break
-          rescue ex : OverflowError | AMQ::Protocol::Error::FrameDecode
-            @log.error { "Could not initialize segment, closing message store: Failed to read segment #{seg} at pos #{mfile.pos}. #{ex}" }
-            close
+            @size += count - acked
+          rescue NoMsgCountError
+            scan_segment(seg, mfile)
           end
+
           mfile.pos = 4
           mfile.unmap # will be mmap on demand
           if is_long_queue
             @log.info { "Loaded #{counter}/#{@segments.size} segments, #{@size} messages" } if (counter &+= 1) % 128 == 0
-          else
-            @log.debug { "Loaded #{counter}/#{@segments.size} segments, #{@size} messages" } if (counter &+= 1) % 128 == 0
           end
-          Fiber.yield
-          @segment_msg_count[seg] = count
+          @log.info { "Loaded #{counter} segments, #{@size} messages" }
+        end
+      end
+
+      private def scan_segment(seg, mfile)
+        mfile.pos = 4
+        count = 0u32
+        loop do
+          pos = mfile.pos
+          ts = IO::ByteFormat::SystemEndian.decode(Int64, mfile.to_slice(pos, 8))
+          if ts.zero?
+            # This means that the rest of the file is zero
+            mfile.resize(pos)
+            break
+          end
+          bytesize = BytesMessage.skip(mfile)
+          count += 1
+          next if deleted?(seg, pos)
+          @bytesize += bytesize
+          @size += 1
+        rescue ex : IO::EOFError
+          break
+        rescue ex : OverflowError | AMQ::Protocol::Error::FrameDecode
+          @log.error { "Could not initialize segment, closing message store: Failed to read segment #{mfile.path} at pos #{mfile.pos}. #{ex}" }
+          close
+        end
+        @segment_msg_count[seg] = count
+
+        unless seg == @segments.last_key # don't append stats to last segment (currently writing to)
+          File.open(mfile.path, "a") do |f|
+            f.truncate(mfile.pos) # delete all potential zeros
+            append_msg_count(f, count)
+          end
+        end
+      end
+
+      private def append_msg_count(file, count)
+        file.write_bytes 0i64 # ts zero
+        file.write_bytes MAGIC
+        file.write_bytes count.to_u32
+        file.write_bytes 0u32 # reserved
```

Comment:
Instead of a reserved integer write a CRC32 of ts, MAGIC and count? Or keep the reserved integer and add a CRC32 field?

Reply:
yes!